db/src/tx.rs

```diff
@@ -695,63 +698,45 @@ impl<'conn, 'a, W> Tx<'conn, 'a, W> where W: TransactWatcher {
         // We need to ensure that callers can't blindly transact entities that haven't been
         // allocated by this store.
 
+        let errors = tx_checking::type_disagreements(&aev_trie);
```

Comment:
I'm a bit surprised why this isn't a side-effect of insertion. After all, you know the attribute at the moment you're finalizing that datom, and so you also know which type the value must be…

Reply:
This is a choice.  As I said (on Slack?) we really want to compute in two different monads: one that exits early and one that collects all errors, for the best consumer experience.  I think it's harder to go from one-error to collecting-errors than the reverse, so I implemented the global one first.